File: aletheia/starters/typescript-agent/agent.ts
/** Separate agent process. Never receives the operator's review credential. */
import { agentClient } from "./transport.js";

const url=process.env.ALETHEIA_URL, token=process.env.ALETHEIA_AGENT_TOKEN;
const namespace=process.env.ALETHEIA_NAMESPACE ?? 'user/demo', action=process.argv[2];
if(!url || !token || !['capture','read'].includes(action ?? '')) {
  throw Error('Run operator_demo.py for a disposable local example, or supply URL, restricted token and capture/read action.');
}
const client=agentClient(url,token);
try {
  const result=await client.GET('/v1/auth/me');
  if(!result.data)throw Error('Principal discovery unavailable.');
  const principal=result.data.data;
  if(!['agent-onboarding-v1','memory-read-v1'].every(p=>principal.supported_profiles.includes(p)))throw Error('Required profiles unavailable; no write attempted.');
  const capabilities=principal.capabilities;
  if(!['memory:read','memory:context','memory:write_candidate'].every(c=>capabilities.some(v=>v===c)) ||
      ['memory:review','memory:write_active','memory:admin'].some(c=>capabilities.some(v=>v===c)))throw Error('Use a restricted agent credential without review, active-write or admin access.');
  if(action==='capture'){
    const key=process.argv[3];
    if(!key)throw Error('Capture requires an explicit operation key as its second argument.');
    const response=await client.POST('/v1/remember',{
      params:{header:{'X-Aletheia-Contract':'agent-onboarding-v1','Idempotency-Key':key}},
      body:{namespace,memory_type:'preference',write_mode:'candidate',subject:'user',predicate:'prefers',
        object:'careful architecture notes',evidence_text:'User prefers careful architecture notes.'}});
    if(!response.data)throw Error('Outcome unknown: retain the same operation key and payload.');
    console.log(JSON.stringify({candidate_id:response.data.data.candidate.id}));
  }else{
    const response=await client.POST('/v1/context-pack',{params:{header:{'X-Aletheia-Contract':'memory-read-v1'}},body:{namespace,query:'architecture',retrieval_mode:'lexical',record_usage:false}});
    if(!response.data)throw Error('Context unavailable.');
    const pack=response.data.data;
    console.log(pack.markdown);
    console.log('Visible context items:',pack.items.length);
    if(pack.items[0]){
      const detail=await client.GET('/v1/claims/{claim_id}/explain',{params:{path:{claim_id:pack.items[0].claim_id},header:{'X-Aletheia-Contract':'memory-read-v1'}}});
      console.log('Provenance:',detail.data?.data.evidence[0]?.content ?? 'No visible evidence');
    }
  }
}catch(error){
  console.error(error instanceof Error ? error.message : 'Memory operation failed; do not blindly repeat a write.');
  process.exitCode=1;
}

File: aletheia/starters/http-agent/agent.py
"""An agent process receives candidate-write access, never review credentials."""
import argparse
import json
import os
from aletheia.client import AletheiaClient, AletheiaClientError


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("action", choices=["capture", "read"])
    parser.add_argument("operation_key", nargs="?")
    args = parser.parse_args()
    token = os.environ.get("ALETHEIA_AGENT_TOKEN")
    url = os.environ.get("ALETHEIA_URL")
    namespace = os.environ.get("ALETHEIA_NAMESPACE", "user/demo")
    if not token or not url:
        raise SystemExit("Set ALETHEIA_URL and ALETHEIA_AGENT_TOKEN, or run operator_demo.py for a disposable local demo.")
    client = AletheiaClient(url, token, timeout=10)
    if not hasattr(client, "current_principal"):
        raise SystemExit("This starter needs the 1.4 development SDK, not the published 1.3.1 SDK.")
    try:
        principal = client.current_principal()
        if not {"memory-read-v1", "agent-onboarding-v1"} <= set(principal.get("supported_profiles", [])):
            raise SystemExit("The service needs memory-read-v1 and agent-onboarding-v1; no candidate was written.")
        capabilities = set(principal["capabilities"])
        needed = {"memory:read", "memory:context", "memory:write_candidate"}
        if not needed <= capabilities or capabilities & {"memory:admin", "memory:review", "memory:write_active"}:
            raise SystemExit("Use a restricted agent token with read/context/write_candidate and without admin, review or active-write access.")
        if args.action == "capture":
            if not args.operation_key:
                raise SystemExit("Capture requires an explicit operation key as its second argument.")
            result = client.remember_candidate(namespace=namespace, memory_type="preference",
                subject="user", predicate="prefers", object="careful architecture notes",
                evidence_text="User prefers careful architecture notes.", idempotency_key=args.operation_key, contract="agent-onboarding-v1")
            print(json.dumps({"candidate_id": result["candidate"]["id"]}))
        else:
            pack = client.context_pack(namespace=namespace, query="architecture", retrieval_mode="lexical", record_usage=False)
            print(pack["markdown"])
            print("Visible context items:", len(pack["items"]))
            if pack["items"]:
                explanation = client.explain_claim(pack["items"][0]["claim_id"])
                print("Provenance:", explanation["evidence"][0]["content"])
    except AletheiaClientError as error:
        raise SystemExit(f"Memory request failed ({error.status_code or 'transport'}). Check discovery, token permissions and namespace with doctor --read-only.") from None
    except (OSError, ValueError):
        raise SystemExit("Memory response unavailable or invalid. Do not blindly repeat a candidate write; ask the operator to inspect pending candidates.") from None


if __name__ == "__main__":
    main()
